envisage/plugins/ipython_kernel/internal_ipkernel.py

```diff
@@ -90,62 +117,51 @@ def init_ipkernel(self, gui_backend):
         # Start IPython kernel with GUI event loop support
         self.ipkernel = gui_kernel(gui_backend)
 
-        # Since ipykernel 4.7, the io_loop attribute of the kernel is not
-        # initialized anymore
-        # Reference: https://github.com/enthought/envisage/issues/107
-        # Workaround: Retrieve the kernel on the IPykernelApp and set the
-        # io_loop without starting it!
-        if NEEDS_IOLOOP_PATCH and not hasattr(self.ipkernel.kernel, 'io_loop'):
-            self.ipkernel.kernel.io_loop = ioloop.IOLoop.instance()
-
         # This application will also act on the shell user namespace
         self.namespace = self.ipkernel.shell.user_ns
         self.namespace.update(dict(self.initial_namespace))
 
     def new_qt_console(self):
         """ Start a new qtconsole connected to our kernel. """
-        console = connect_qtconsole(self.ipkernel.connection_file,
-                                    argv=['--no-confirm-exit'])
+        console = ipykernel.connect.connect_qtconsole(
+            self.ipkernel.connection_file,
+            argv=['--no-confirm-exit'],
+        )
         self.consoles.append(console)
         return console
 
     def cleanup_consoles(self):
         """ Kill all existing consoles. """
         for c in self.consoles:
             c.kill()
+            c.wait()
             c.stdout.close()
             c.stderr.close()
         self.consoles = []
 
     def shutdown(self):
-        """ Shutdown the kernel.
+        """ Shut the kernel down.
 
         Existing IPython consoles are killed first.
         """
         if self.ipkernel is not None:
             self.cleanup_consoles()
-            self.ipkernel.shell.exit_now = True
-            self.ipkernel.cleanup_connection_file()
-
-            # The stdout and stderr streams created by the kernel use the
-            # IOPubThread, so we need to close them and restore the originals
-            # before we shut down the thread. Without this, we get obscure
-            # errors of the form "TypeError: heap argument must be a list".
-            kernel_stdout = sys.stdout
-            if kernel_stdout is not self._original_stdout:
-                sys.stdout = self._original_stdout
-                kernel_stdout.close()
-
-            kernel_stderr = sys.stderr
-            if kernel_stderr is not self._original_stderr:
-                sys.stderr = self._original_stderr
-                kernel_stderr.close()
-
-            self.ipkernel.iopub_thread.stop()
+            self.ipkernel.shutdown()
```

Comment:
Can you comment on how `ipkernel.shutdown` does the job of 

```
self.ipkernel.shell.exit_now = True
self.ipkernel.cleanup_connection_file()
self.ipkernel.iopub_thread.stop()
```

?

Reply:
Or maybe the removed code isn't relevant anylonger or was the source of bugs.

Reply:
The `cleanup_connection_file` and `iopub_thread.stop` are still happening: see the `kernelapp` code. The code's just been moved to a more appropriate place (it's the `kernelapp` that sets this up, so it's the kernelapp's job to shut them down; we couldn't do that before because we were trying to work with the upstream kernelapp).

The `exit_now` code should never have been there, as discussed above.